changefeedccl: refactor to stop passing around untyped string maps

HonoreDB opened the pull request:
Why did I think this would be a quick win. This was slow and boring
and might be too big a diff to do in one go.

This PR makes the WITH options map into a typed object,
and moves a lot of the parsing and validation code into methods on
that object so that sinks, encoders, and so on can take small
typed objects instead of the whole giant untyped map. The only
time we still treat it as a giant map is input validation.

I've kept the parsing of JSON WITH options in the individual sinks
to go along with the 'quick and dirty add anything' vibe they have.

There's still work to do here: the protos used to communicate
state live outside of ccl, so we're still passing around the
untyped map a lot. And some more logic that could be moved into
options.go . But the diff is big enough already.

Release note: None

Closes https://github.com/cockroachdb/cockroach/issues/80105

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/80247)
<!-- Reviewable:end -->

HonoreDB committed:
changefeedccl: refactor to stop passing around untyped string maps

Why did I think this would be a quick win. This was slow and boring
and might be too big a diff to do in one go.

This PR makes the WITH options map into a typed object,
and moves a lot of the parsing and validation code into methods on
that object so that sinks, encoders, and so on can take small
typed objects instead of the whole giant untyped map. The only
time we still treat it as a giant map is input validation.

I've kept the parsing of JSON WITH options in the individual sinks
to go along with the 'quick and dirty add anything' vibe they have.

There's still work to do here: the protos used to communicate
state live outside of ccl, so we're still passing around the
untyped map a lot. And some more logic that could be moved into
options.go . But the diff is big enough already.

Release note: None

miretskiy reviewed (commented):
Initial set of comments -- partial review.

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80247)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @miretskiy)

___
*[`pkg/ccl/changefeedccl/changefeedbase/errors.go` line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjp-r1-118:-N1FMzrm0Z28R99Jr_wd:b-nfu5q1) ([raw file](https://github.com/cockroachdb/cockroach/blob/257b3ae3c4491bde21f9ec4d2db4f756ec39584b/pkg/ccl/changefeedccl/changefeedbase/errors.go#L118)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Feels like this is not necessarily options specific change? Could it be pulled into separate change? Could be part of the same pull request; just logically feels... different.
</blockquote></details>

This was part of removing the sql dependency from changefeedbase. Which is no longer necessary since I'm getting rid of the jobspb dependency onto changefeedbase, but given that I've done it I'd rather not undo it. I'll move it into a separate commit before merge.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r1-196:-N1FQPUbAY9hbrx7OdSX:b-6e0wkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/257b3ae3c4491bde21f9ec4d2db4f756ec39584b/pkg/ccl/changefeedccl/changefeedbase/options.go#L196)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

As your pr description says: this is probably a lot more effort than originally thought.
👏 for taking this on.

I was wondering if, perhaps, this code can be somewhat simplified.  Take a look at
https://gist.github.com/miretskiy/af25259bd29ddace5b7402acd7308468

I haven't tested/tried this gist, so I don't know if this might be helpful.  Basically, I added a `StringOptionsAsCustomConfig` method -- this method is almost 1-1 copy of the "old" `TypeAsStringOpts`.

The idea is that we have an interface that defines required values for the option, but also let's you define custom `FromString` method.

The way that this might work (or at least I think it might work), is that we have this mega config struct -- basically what you have (`EncodingOptions`, etc).
So, I think you could have that struct implement `KVCustomConfig` interface to provide custom parsing for e.g. enums, etc.

I don't know if this gist would help -- if not -- my apologies; just want to see what you think about something like that.
</blockquote></details>

I definitely think this code could be simplified using something like this that takes custom validations. I hadn't thought of replacing the `TypeAsStringOpts` method entirely, that's a cool idea. I was surprised we didn't seem to have things like TypeAsInt.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 843 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r1-843:-N1FOmlE-y_bfCbuJsBg:barsoqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/257b3ae3c4491bde21f9ec4d2db4f756ec39584b/pkg/ccl/changefeedccl/changefeedbase/options.go#L843)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Is this a bit too fancy? `case 1` seems to be strange -- why would you have an enum that allows exactly 1 value? it's a flag.
And the rest... I don't know strings.Join seems sufficient; possibly wrapped in `()` or `[]`.
</blockquote></details>

We have a single-value enum right now, `compression` only takes the value `gzip` because that's the only compression format we support. The rest of the fanciness is to avoid changing existing error messages.
___
*[`pkg/sql/delegate/show_changefeed_jobs.go` line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N1AnSdcOwSw-RwbUnjf-r1-66:-N1FPywH8nktCnX-bnI6:b-61cp36) ([raw file](https://github.com/cockroachdb/cockroach/blob/257b3ae3c4491bde21f9ec4d2db4f756ec39584b/pkg/sql/delegate/show_changefeed_jobs.go#L66)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Any parse tests need to be updated?
</blockquote></details>

Not according to CI I think. Is there something that generally needs to be updated when you update a SQL statement like this?
___
*[`pkg/jobs/jobspb/BUILD.bazel` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_JO2rIUrcCgb4s-r1-49:-N1FPXp-7xPO-unfYJt8:bboeq8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/257b3ae3c4491bde21f9ec4d2db4f756ec39584b/pkg/jobs/jobspb/BUILD.bazel#L49)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

why? Also... it kinda makes me sad that we have to add job implementation specific dependency.
</blockquote></details>

Oops, I got rid of this dependency but forgot I needed to manually update this file because of the `# keep` comment. This was for gogoproto but we're trying not to add more dependencies from non-ccl onto ccl. Removed.


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80247)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @miretskiy)

___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r1-196:-N3-f7kFCnHIJq7nVHg_:bqaogbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/257b3ae3c4491bde21f9ec4d2db4f756ec39584b/pkg/ccl/changefeedccl/changefeedbase/options.go#L196)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I definitely think this code could be simplified using something like this that takes custom validations. I hadn't thought of replacing the `TypeAsStringOpts` method entirely, that's a cool idea. I was surprised we didn't seem to have things like TypeAsInt.
</blockquote></details>

Do you mind actually if I wait to do this until I'm allowed to use generics? I'd love to have the method just cast everything appropriately and it'll be a lot cleaner when I can.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 205 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r1-205:-N3-fr-z1JInNGqXs5dI:b2k4tvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/257b3ae3c4491bde21f9ec4d2db4f756ec39584b/pkg/ccl/changefeedccl/changefeedbase/options.go#L205)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

would OptionTypeDuration be a better name?
</blockquote></details>

Yup, changed.
___
*[`pkg/ccl/changefeedccl/changefeedbase/errors.go` line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjp-r1-118:-N3-dT2L8NleOspXv4XD:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/257b3ae3c4491bde21f9ec4d2db4f756ec39584b/pkg/ccl/changefeedccl/changefeedbase/errors.go#L118)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

This was part of removing the sql dependency from changefeedbase. Which is no longer necessary since I'm getting rid of the jobspb dependency onto changefeedbase, but given that I've done it I'd rather not undo it. I'll move it into a separate commit before merge.
</blockquote></details>

Removed.


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -243,11 +319,14 @@ var CloudStorageValidOptions = makeStringSet(OptCompression)
 
 // CaseInsensitiveOpts options which supports case Insensitive value
-var CaseInsensitiveOpts = makeStringSet(OptFormat, OptEnvelope, OptCompression, OptSchemaChangeEvents, OptSchemaChangePolicy, OptOnError)
+var CaseInsensitiveOpts = makeStringSet(OptFormat, OptEnvelope, OptCompression, OptSchemaChangeEvents, OptSchemaChangePolicy, OptOnError, OptInitialScan)
```
nit: better wrapping?

miretskiy commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -299,3 +378,512 @@ var VersionGateOptions = map[string]clusterversion.Key{
+
+// StartFromCursor returns true if we're starting from a
+// user-provided timestamp.
+func (s StatementOptions) StartFromCursor() bool {
```
nit: HasStartCursor?

miretskiy commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -299,3 +378,512 @@ var VersionGateOptions = map[string]clusterversion.Key{
+	MVCCTimestamps          bool
+	Diff                    bool
+	AvroSchemaPrefix        string
+	ConfluentSchemaRegistry string
```
SchemaRegistryURI?

miretskiy commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -299,3 +378,512 @@ var VersionGateOptions = map[string]clusterversion.Key{
+// ClientTimeout is nil if not set as the default
+// is different from 0.
+type WebhookSinkOptions struct {
+	JSON          string
```
JSON is the config for webhook? Perhaps call it config?

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80247)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @miretskiy)

___
*[`pkg/ccl/changefeedccl/encoder.go` line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjr-r4-56:-N35mCwmDzV_SsbdxZ2g:b-bokp79) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/encoder.go#L56)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

should we perhaps use something other than assertion failed? Maybe pgerror?
</blockquote></details>

It's assertion failed because it's already been converted into a FormatType so this should be unreachable code.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 645 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjy-r4-645:-N35neFsC8GKOto1cQCF:bn5tg1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/sink_kafka.go#L645)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

other sink (e.g. sink_cloudstorage) takes a struct -- which is nicer than the string...
Perhaps passing statement options might not be such a bad thing (and  of course, it's fine to pass in GetKafkaJSON... string to buildKafkConfigBelow?
</blockquote></details>

Why is a struct nicer than a string here?
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 211 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-211:-N3631aw4ZXTpJjuLugx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L211)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This file is now very big.  And it has code.  And now it really wants to have a test friend :)
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 231 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-231:-N35nosk1s0__BRCWCxf:b-2mgc5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L231)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

So, this option cannot be combined with CanBeEmpty, ie.. have a single option
DefaultIfEmpty -- which if not empty sets the default when option is empty?
</blockquote></details>

The problem is that leaves no way to distinguish between `orEmptyMeans("")` and can't be empty.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 276 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-276:-N35oH_-Baj5kKBqdWms:b-bxoy1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L276)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

this is the only place where thatCanBeZero is used -- do we need it?
</blockquote></details>

(here and the next line). I'm trying to make this code block maximally human-readable so we can link to it as a substitute for auto-generated docs. But can take it out if you'd rather.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 326 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-326:-N35oldZBQlTO171LCCH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L326)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: better wrapping?
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 384 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-384:-N35qPDUCBKVvL2FtyNN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L384)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

If I'm not mistaken, StatementOptions is now a main type in this file -- callers construct StatementOptions and operate on that thereafter.  I think it needs to be moved to the top of this file.  Perhaps with somewhat expand comment to convey this types importance.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 397 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-397:-N35qRPeAkvMDcXiXZPQ:b5qy7aj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L397)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm concerned about the lazy nature of StatementOptions.
Consider how previous map was used; it was fairly common to access this map many times, sometimes to even check the same option.  It's fine, most likely.
But now, if I want e.g. EncodingOptions, I have to call GetEncodingOptions() -- and that method does many things, including multiple access to that map, as well as validations.

I think that's a bit unfortunate because the caller now must be aware of that. 
What do you think about changing StatementOptions not to be lazy, so that it becomes a full fledged struct with all others options parsed and initialized when constructing StatementOptions?
```
type StatementOptions struct {
  EncodingOptions
  TableTolerances
   ...
}
```

</blockquote></details>

There is a reason I didn't do it that way...not a great reason, but a reason, which is that our existing create changefeed flow relies on different kinds of errors being raised at different points. If I create everything eagerly, I can't validate as I go because of things like ForceKeyInValue not having been called yet. This is definitely fixable but maybe should be a TODO. 
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 446 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-446:-N35rP8n3uRAEbT3KfEk:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L446)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: I would just renamed `ForEachWithRedaction`
</blockquote></details>

done
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 458 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-458:-N35rdc1CLVsbLytTJNc:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L458)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: HasStartCursor?
</blockquote></details>

done
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 469 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-469:-N35rcjY082TZ5Wb2SZ8:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L469)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Perhaps HasEndTime?
</blockquote></details>

done
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 581 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-581:-N35rnEJ2Ag5aAJX-loN:b-fnr9uu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L581)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Perhaps this is a bit too much in terms of decomposition?
Maybe a simple `func ShouldQualifyNames() bool` would do? And if you dislike
boolean here, you could 
```
type NameQualification bool
const FullyQualified NameQualification = true
```

</blockquote></details>

I wanted to make it clear that this is talking about statement time names and not anything else...and we'll probably need more attributes here in the future, sadly.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 590 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-590:-N35s7aN29NvHmKrw6mD:b-hrxtyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L590)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: `TableHandling` as a name for the struct, and then you can probably drop `Can` prefixes from the fields?
</blockquote></details>

I'd like to distinguish this from config that says *how* to handle unusual tables.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 617 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-617:-N35tl_z6kWd6N11UD1F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L617)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

SchemaRegistryURI?
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeedvalidators/options_sql_validator.go` line 21 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06mijkl8Th5LXuYCih-r4-21:-N35tyMV9bNId8RNOLVw:bqzq28l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedvalidators/options_sql_validator.go#L21)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I wonder if it's better to move all of this code directly into changefeedbase.
I guess the choice here is between arguably nice code location for validations, and on the other hand, the need to expose things like CanBeEmpty fields -- which I believe could have stayed private.
</blockquote></details>

This package separation was to remove the sql dependency from changefeedbase, which was causing circular dependencies at one point.
___
*[`pkg/ccl/changefeedccl/encoder_test.go` line 53 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjs-r4-53:-N35uQtnAMc9BgmFnlZN:b9iiy75) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/encoder_test.go#L53)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Do you need to specify UpdatedTimestamps & Diff?
</blockquote></details>

Just for readability, it's part of a test truth table.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80247)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @miretskiy)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 645 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjy-r4-645:-N364QFi06m9mM13fuTY:b3bkm76) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/sink_kafka.go#L645)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Why is a struct nicer than a string here?
</blockquote></details>

Consistency (with other sink types?) What does that string represent? I thought we're trying to get away from passing bags of strings -- I guess a single stirng is an improvement...  But still, kinda feels we are passing sink specific configuration.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 397 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-397:-N363iewE7aabcIEoaDa:bibtbl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L397)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

There is a reason I didn't do it that way...not a great reason, but a reason, which is that our existing create changefeed flow relies on different kinds of errors being raised at different points. If I create everything eagerly, I can't validate as I go because of things like ForceKeyInValue not having been called yet. This is definitely fixable but maybe should be a TODO. 
</blockquote></details>

I understand; however, I think my concern is quite real:
```
if so.GetDecodingOptions().Foo {
  ....
}

.... 
if so.GetDecodingOptions().Bar {
}
....
```

At the very least you could memoize things?
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 581 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-581:-N363zSB3jrkQ4r6LgNE:b-5x9j5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L581)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I wanted to make it clear that this is talking about statement time names and not anything else...and we'll probably need more attributes here in the future, sadly.
</blockquote></details>

Quite possible... But for now, maybe follow the rule of not adding unless you need it?
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 590 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-N06iq_ICbNg3ndOjvjq-r4-590:-N36451b1k4mfEGY1mzY:bsjs2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced1c2f67da8142363e5770edb70415f1ff1d7db/pkg/ccl/changefeedccl/changefeedbase/options.go#L590)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I'd like to distinguish this from config that says *how* to handle unusual tables.
</blockquote></details>

I just find TableTolerances talking about column families and virtual columns to be confusing on both table, and tolerances counts :)
I don't know.. Just feels like maybe there is a better name out there.


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (commented):
All comments addressed. If there's anything more you'd like to change, I'd rather come back from vacation to find that you stole this PR and merged it yourself than to find that I have to rebase it and fix more merge conflicts. (Or @surajr10 might be able to adopt this, he's working on an options issue anyway).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80247#-:-N3fSd0r0yA6hGFjpqFV:bh0msf1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @miretskiy)



<!-- Sent from Reviewable.io -->
